### PR TITLE
[main][feat] Add EPS optimal stable swap strategy 

### PR DESCRIPTION
**Pull request by alpaca-junior**
## Description
- The strategy will calculate the optimal swap amount by using Multidimensional Newton’s Method and find out path for swapping with Ellipsis Stable Swap pool given the amounts of two underlying tokens. The result of the optimal swap should provide the optimal swap path and amounts of those two tokens for adding liquidity with constant product AMM with least price slippage and most optimized capital efficiency.
- The strategy will handle the calculation against Base Pool and Meta Pool of Ellipsis (see https://resources.curve.fi/faq/base-and-metapools).

## Why?
Optimized capital efficiency for stable pairs.

### CODE REVIEW

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+    PRECISION = 10**18;
+
+    // 2. Assign Pancakeswap dependency contracts
+    router = IPancakeRouter02(_router);
+    factory = IPancakeFactory(router.factory());
+
+    // 3. Assign EPS depency contracts & variables
+    epsPool = ICurveBase(_epsPool);
+    epsA = epsPool.A();
+    epsFee = epsPool.fee();
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:73]: `epsPool.A()` and `epsPool.fee()` can be changed over time. Please observe it and update when strategy execute too.

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+    uint256 D;
+    uint256 S;
+    uint256 U;
+    uint256 Xfp;
+    uint256 Yfp;
+    uint256 N_COINS;
+    uint256 Ann;
+  }
+
+  function execute(
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:110]: Please add description over each function some thing like:
```
/// @dev Execute EPS strategy. Received stables from Worker, optimal swap on EPS, return LP.
/// @param ...
/// @param ...
```

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    (uint256 farmingTokenAmount, uint256 minLPAmount) = abi.decode(data, (uint256, uint256));
+
+    IWorker worker = IWorker(msg.sender);
+    address baseToken = worker.baseToken();
+    address farmingToken = worker.farmingToken();
+    require(espPoolTokens[baseToken] && espPoolTokens[farmingToken], "!worker not compatiable");
+
+    vault.requestFunds(farmingToken, farmingTokenAmount);
+    uint256 baseTokenAmount = baseToken.myBalance();
+    BalancePair memory user_balances = BalancePair(baseTokenAmount, farmingTokenAmount);
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:124]: Please use camelCase for variables

File: .nvmrc [hunk 4]
```diff
@@ -1 +1 @@
-12.16.2
\ No newline at end of file
+14.15.4
```
Comment (almost-llamas) [inline .nvmrc:1]: Not sure it's safe to change this. Can you confirm that all unit tests are fine?

Comment (spicysquid168) [reply .nvmrc:1]: Should be fine

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+contract StrategyAddStableOptimal is IStrategy, OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe {
+  using SafeMath for uint256;
+  using SafeToken for address;
+
+  /* ========== STATE VARIABLES ========== */
+  IPancakeRouter02 public router;
+  IPancakeFactory public factory;
+
+  ICurveBase public epsPool;
```
Comment (ricardo-alpaca) [inline contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:44]: Question: Usually we will make `Strategy` to be pair-agnostic by retrieving the `baseToken` and `farmingToken` from the calling worker and derive the pair address. I know we cannot do that with EPS, but by fixing a single `Strategy` contract to handle only one EPS pool would mean we will have to deploy multiple `Strategy` contracts for each pool we are going to use. We will need 3 pools which are 3eps, TUSD and DAI to cover all of our stablecoin farms.

Should we do it in a more dynamic fashion?

A solution I could think of is having one `Strategy` only and we will declare a mapping to keep the pool addresses which are indexed by the token addresses inside those pools like what `PancakeFactory` did.

@spicysquid168 @alpaca-junior

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+    IWorker worker = IWorker(msg.sender);
+    address baseToken = worker.baseToken();
+    address farmingToken = worker.farmingToken();
+    require(espPoolTokens[baseToken] && espPoolTokens[farmingToken], "!worker not compatiable");
+
+    vault.requestFunds(farmingToken, farmingTokenAmount);
+    uint256 baseTokenAmount = baseToken.myBalance();
+    BalancePair memory user_balances = BalancePair(baseTokenAmount, farmingTokenAmount);
+
+    PoolState memory fp_state;
```
Comment (ricardo-alpaca) [inline contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:126]: Could we change `fp` and `sb` prefixes into something readable? I'm guessing `fp` is `farmingPool`.

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 2 above)
Comment (alpaca-junior) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:112]: Add another refresh at the beninging ... _wait_ .. the beginning of execute( ).

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 1 above)
Comment (alpaca-junior) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:73]: Added another refresh at the beninging ... _wait_ .. the beginning of execute( ).

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 5 above)
Comment (spicysquid168) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:44]: That can be done too.  Having something like `mapping(address => mapping(address => address)) epsPools` should solve that.

**Comment by spicysquid168** [comment]
I attached the test result of add stable optimal here:
[EPS-add-stable-optimal.csv](https://github.com/alpaca-finance/bsc-alpaca-contract/files/7117119/EPS-add-stable-optimal.csv)

From all 714 test-cases with various underlying pools, the max debris (leftover from optimal swap due to calculation) is like 6 wei.

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 6 above)
Comment (alpaca-junior) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:132]: I added comments before each function and that might help clarifying those variables. If they are still ambiguous, can you suggest one?

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 2 above)
Comment (alpaca-junior) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:112]: updated

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 3 above)
Comment (alpaca-junior) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol]: updated all. except for `i_sb, j_sb, i_fp, j_fp` I think it's more readable this way e.g. `jSb` seems confusing as indexing variable.

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 6 above)
Comment (ricardo-alpaca) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol]: That should do it.

File: contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol (hunk 5 above)
Comment (ricardo-alpaca) [reply contracts/6/protocol/strategies/eps/StrategyAddStableOptimal.sol:44]: This comment is obsoleted now. As Ellipsis only has one Base Pool for stablecoins which is 3EPS. I don't think we need to make it dynamic. We will do Meta Pool in another smart contract.

**Review by josetommaso — DISMISSED** [review]

**Pull request closed without merging**